Handle locked collections/items

camh- opened the pull request:
Handle get/store/erase on locked collections and items. When an item is 
locked, we typically get a "prompt" dbus object that we need to call the
"Prompt" method on. This causes the Secret Service (gnome-keyring) to pop
up a window prompting the user for the password to unlock the item.

Use a Go 1.23 range/iterator function to iterate over the search results 
when looking up a secret. This handles the unlocked and locked results 
transparently, unlocking any matching locked items found. The iterator also
returns only the items that match the attributes exactly. This makes the
implementation of Get and Delete simpler.

Upgrade the hermit verrsion of Go and bump the "go" version in `go.mod`
so that we can use range functions.

camh- committed:
hermit: Bump Go to 1.23, golangci-lint to 1.60.3

Bump the Go version to Go 1.23 so we can use range functions.

Bump golangci-lint to handle the new version of Go. Disable
`exportloopref` linter as it is deprecated. Disable the `gci` linter
because it is emitting nonsensical errors and I don't use `gci`.
Ignore one new lint bogus warning.

    hermit upgrade go golangci-lint

camh- committed:
Handle locked collections/items

Handle get/store/erase on locked collections and items. When an item is
locked, we typically get a "prompt" dbus object that we need to call the
"Prompt" method on. This causes the Secret Service (gnome-keyring) to
pop up a window prompting the user for the password to unlock the item.

Use a Go 1.23 range/iterator function to iterate over the search results
when looking up a secret. This handles the unlocked and locked results
transparently, unlocking any matching locked items found. The iterator
also returns only the items that match the attributes exactly. This
makes the implementation of Get and Delete simpler.

Bump the "go" version in `go.mod` so that we can use range functions.

juliaogris commented on dbus.go:
```diff
@@ -236,13 +221,72 @@ func (ss *SecretService) Delete(attrs map[string]string, expectedPassword string
+			if !ok && err == nil {
+				continue
+			}
+			itemPath, err = ss.unlockItem(itemPath)
```
potentially overrides non-nil error from before

camh- replied:
Thanks for finding that. Fixed.

juliaogris commented on dbus.go:
```diff
@@ -155,7 +142,23 @@ func (ss *SecretService) Store(label string, attrs map[string]string, secret str
+		return err
+	}
+
+	if promptPath != dbus.ObjectPath("/") {
```
maybe factor out `dbus.ObjectPath("/")` as const `noPrompt`

camh- replied:
I've done this in all places. I think it does read better.

juliaogris commented on dbus.go:
```diff
@@ -155,7 +142,23 @@ func (ss *SecretService) Store(label string, attrs map[string]string, secret str
+		ok, err := ss.prompt(promptPath)
+		if err != nil {
+			return err
+		} else if !ok {
```
or maybe if on the next line?

camh- replied:
I've gotten rid of this altogether by returning the cancelled bool as an error instead.

juliaogris commented on dbus.go:
```diff
@@ -288,3 +332,74 @@ func (ss *SecretService) attrsMatch(attrs map[string]string, itemPath dbus.Objec
 	return true, nil
 }
+
+func (ss *SecretService) unlockItem(itemPath dbus.ObjectPath) (dbus.ObjectPath, error) {
```
Rework as `unlockObject` and use in `ss.Store(...)`

camh- replied:
Done.

juliaogris commented on dbus.go:
```diff
@@ -288,3 +332,74 @@ func (ss *SecretService) attrsMatch(attrs map[string]string, itemPath dbus.Objec
+
+	if ok, err := ss.prompt(promptPath); err != nil {
+		return "", err
+	} else if !ok {
```
no need for else if

juliaogris commented on dbus.go:
```diff
@@ -236,13 +221,72 @@ func (ss *SecretService) Delete(attrs map[string]string, expectedPassword string
 
-	if promptPath != dbus.ObjectPath("/") {
-		fmt.Fprintln(os.Stderr, "TODO: Got prompt on delete. Sorry, can't do that yet")
+	if promptPath == dbus.ObjectPath("/") {
```
use noPrompt ?

juliaogris commented on dbus.go:
```diff
@@ -236,13 +221,72 @@ func (ss *SecretService) Delete(attrs map[string]string, expectedPassword string
+		return nil
+	}
+
+	ok, err := ss.prompt(promptPath)
```
maybe wrap up ok&err into an error as you use it in the same way in a couple of locations.

juliaogris commented on dbus.go:
```diff
@@ -288,3 +332,74 @@ func (ss *SecretService) attrsMatch(attrs map[string]string, itemPath dbus.Objec
+		}
+		break
+	}
+	return !cancelled, nil
```
return Sentinel for cancel (will cleanup all the if-elses as well).

camh- replied:
Done. As you say, this has cleaned up the prompt handling in all places it was called. I also added one more place where I should be prompting, although gnome-keyring does not ever return a prompt there, it is more complete in case another secret service is used.

juliaogris commented on dbus.go:
```diff
@@ -288,3 +332,74 @@ func (ss *SecretService) attrsMatch(attrs map[string]string, itemPath dbus.Objec
+}
+
+// unlock attempts to unlock the objects given and returns the paths for the
+// objects that were unlocked and a prompt path to unlock the remainder.
```
Add link to docs (below / above as well)

camh- replied:
Done for unlocking and prompting - hopefully this gives some background to help understand these operations.

juliaogris reviewed (approved):
thanks for explaining iterators :green_heart:

camh- commented:
@juliaogris PTAL if you like. I've addressed all your comments - code reads nicer now, thanks. I've left the individual fixup commits here for now (although I didn't bother to link them to the comments - a bit tedious). It should help see how I made the changes.
